### DISCUSSION TITLE
Types for `data: (context, event) =>` are 'any'

### DISCUSSION THREAD

**Discussion by lukehorvat** [statelyai/xstate#1261] [General] [answered]
Hi. I have the following `foo` state that invokes a child machine with context determined by an event:

```tsx
foo: {
  invoke: {
    src: ChildMachine,
    data: (context, event) => ({
      mode: event.type === 'ONE' ? 1 : event.type === 'TWO' ? 2 : undefined,
    }),
    onDone: 'bar',
  },
},
```

The `(context, event) =>` part gives a TS error:

> Parameter 'context' implicitly has an 'any' type.
> Parameter 'event' implicitly has an 'any' type.

Why doesn't it know the parameter types?

I fixed the error for now by being more explicit, but it doesn't feel nice:

```tsx
foo: {
  invoke: {
    src: ChildMachine,
    data: (context: ParentContext, event: ParentEvents): Partial<ChildContext> => ({
      mode: event.type === 'ONE' ? 1 : event.type === 'TWO' ? 2 : undefined,
    }),
    onDone: 'bar',
  },
},
```

Side note: is it possible to define `data` the serialized way?

**Comment by Andarist** [comment, marked as answer]
Thank you for the report - I've started working on fixing this. The problem is that currently `any` is included in some union types and this always makes the whole union equal to `any`.

> Side note: is it possible to define data the serialized way?

You could use `services` to do that though - you would just have to return `ChildMachine.withContext(/* ... */)` from a service creator based on received arguments (`context` and `event`).

**Reply by lukehorvat** [reply]
Just tried v4.11.0. Works great. Thanks!
